pkg/cloud/azure/services/networkinterfaces/networkinterfaces.go

```diff
@@ -234,3 +254,29 @@ func (s *Service) Delete(ctx context.Context, spec azure.Spec) error {
 	klog.V(2).Infof("successfully deleted nic %s", nicSpec.Name)
 	return err
 }
+
+func subnetHasIPv6(subnet network.Subnet) bool {
```

Comment:
I don't know if this is the right thing, but I don't see any other way to tell if a subnet is IPv6-enabled.

Another option is to make the decision on whether to configure an IPv6 address explicit in the provider config. If the subnet doesn't have IPv6 address space, creating the NIC will fail and the Machine will go into a failed state.